content/advent-2017/messaging-framework.md

```diff
@@ -0,0 +1,709 @@
+		case nil:
+			// new data available
+			c.onDataEvent(c, buf[:n])
+		default:
+			log.Fatalf("Receive data failed:%s", err)
+			c.conn.Close()
+			c.onConnectionEvent(c, CONNECTION_EVENT_TYPE_CONNECTION_GENERAL_ERROR, err)
+			return
+		}
+	}
+}
+```
+
+## Kafka Consumer
+
+Its role is to consume messages from our Kafka broker, and to broadcast them back to relevant clients by their uids.  
+In this example we are consuming from multiple topics using the [cluster implementation of sarama](github.com/bsm/sarama-cluster).
+
+let's define our ```Consumer``` struct:  
+```go
+type Consumer struct {
+	consumer *cluster.Consumer
+	callbacks ConsumerCallbacks
+}
+```
+```ConsumerCallbacks``` are:
+```go
+
+type ConsumerCallbacks struct {
+	OnDataReceived func(msg []byte)
+	OnError func(err error)
+}
+```
+The constructor receives the callbacks and relevant details to connect to the topic:
+```go
+func NewConsumer(callbacks ConsumerCallbacks,brokerList []string, groupId string, topics []string) *Consumer {
+	consumer := Consumer{callbacks:callbacks}
+
+	config := cluster.NewConfig()
+	config.ClientID = uuid.NewV4().String()
+	config.Consumer.Offsets.Initial = sarama.OffsetNewest
+	saramaConsumer, err := cluster.NewConsumer(brokerList, groupId, topics, config)
+	if err != nil {
+		panic(err)
+	}
+	consumer.consumer = saramaConsumer
+	return &consumer
+
+}
+```
+It will consume permanently on a new goroutine:
+```go
+func (c *Consumer) Consume() {
+	// Create signal channel
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+
+	// Consume all channels, wait for signal to exit
+	go func(){
+		for {
+			select {
+			case msg, more := <-c.consumer.Messages():
+				if more {
+					if c.callbacks.OnDataReceived!=nil {
+						c.callbacks.OnDataReceived(msg.Value)
+					}
+					fmt.Fprintf(os.Stdout, "%s/%d/%d\t%s\n", msg.Topic, msg.Partition, msg.Offset, msg.Value)
+					c.consumer.MarkOffset(msg, "")
+				}
+			case ntf, more := <-c.consumer.Notifications():
+				if more {
+					log.Printf("Rebalanced: %+v\n", ntf)
+				}
+			case err, more := <-c.consumer.Errors():
+				if more {
+					if c.callbacks.OnError!=nil {
+						c.callbacks.OnError(err)
+					}
+				}
+			case <-sigchan:
+				return
+			}
+		}
+	}()
+
+}
+```
+
+## Kafka Producer
+Its role is to produce messages to our Kafka broker.  
+In this example we are producing to a single topic.  
+This section is mainly inspired from the example in https://github.com/Shopify/sarama/blob/master/examples/http_server/http_server.go
+
+Let's define our ```Producer``` Struct:
+```go
+type Producer struct {
+	asyncProducer sarama.AsyncProducer
+	callbacks     ProducerCallbacks
+	topic         string
+}
+```
+These are the callbacks:
+```go
+type ProducerCallbacks struct {
+	OnError func(error)
+}
+```
+```Producer``` is constructed with the callbacks for error, and the details to connect to the Kafka broker including optional ssl configurations:
```

Comment:
```Producer``` will write it a new line, iirc. Maybe you can use `Producer`?

Reply:
Thanks, will replace that globally throughout the article